### PR TITLE
enable bf16 atomic add

### PR DESCRIPTION
**Pull request by zhuhaozhe**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96472

This PR just enabled BF16 atomic add to solve functionality issue for https://github.com/pytorch/pytorch/issues/95197 and https://github.com/pytorch/pytorch/issues/95189. `atomic add`are following https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/AtomicAddFloat.h. The old one with `volatile` keyword and`weak exchange` is hard to support BF16 since BF16 is not a dtype and do not support `bfloat16 val = *addr` while `addr` is `volatile bfloat16 *`

But I do not know whether we should use BF16 atomic add since it accuracy is low. It suffered from:
 - non deterministic parallel reduction (FP32 also suffered but low precision suffered more. https://github.com/pytorch/pytorch/issues/93582
 -  Do no using accumulation type.

From my understanding, it is hard to have high accuracy if we perform add to a `bfloat16 ptr` in `store`
```
extern "C" void kernel(const bfloat16* __restrict__ in_ptr0,
                       const long* __restrict__ in_ptr1,
                       const bfloat16* __restrict__ in_ptr2,
                       bfloat16* __restrict__ out_ptr0)
{
    #pragma omp parallel num_threads(64)
    {
        {
            #pragma omp for 
            for(long i0=0; i0<1254400; i0+=1)
            {
                auto tmp0 = static_cast<float>(in_ptr0[i0]);
                out_ptr0[i0] = tmp0;
            }
        }
        {
            #pragma omp for 
            for(long i0=0; i0<600; i0+=1)
            {
                #pragma GCC ivdep
                for(long i1=0; i1<12544; i1+=1)
                {
                    auto tmp0 = in_ptr1[i0];
                    auto tmp1 = static_cast<float>(in_ptr2[i1 + (12544*i0)]);
                    atomic_add(&out_ptr0[i1 + (12544*tmp0)], static_cast<bfloat16>(tmp1));
                }
            }
        }
    }
}
''')
```
Using local buffer for accumulation and using single thread to add local buffer together might be a better solution for both two issues, but may have impact on performance. Do you have some suggestions are this? CC @EikanWang, @jgong5.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @soumith @anijain2305 @desertfire

### CODE REVIEW

File: torch/_inductor/codegen/cpp_prefix.h [hunk 1]
```diff
@@ -38,24 +44,44 @@ float randn_cpu(uint32_t seed, uint32_t offset) {
...
-  } while (!atomic_addr->compare_exchange_weak(expected, desired,
-                                               std::memory_order_relaxed));
+  alt_type* expected_intV = (alt_type*)(&expected.intV);
+  while (!std::atomic_compare_exchange_strong(
+      atomic_addr, expected_intV, desired.intV)) {
+#ifdef __aarch64__
+    __asm__ __volatile__("yield;" : : : "memory");
+#else
+    _mm_pause();
+#endif
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_prefix.h:81]: why do we need this?

File: torch/_inductor/codegen/cpp_prefix.h [hunk 2]
```diff
@@ -38,24 +44,44 @@ float randn_cpu(uint32_t seed, uint32_t offset) {
...
 
-  std::atomic<alt_type> *atomic_addr = (std::atomic<alt_type> *)addr;
-  do {
-    T val = *addr;
-    reinterpret_cast<T *>(&expected)[0] = val;
-    reinterpret_cast<T *>(&desired)[0] = val + offset;
-  } while (!atomic_addr->compare_exchange_weak(expected, desired,
-                                               std::memory_order_relaxed));
+  alt_type* expected_intV = (alt_type*)(&expected.intV);
+  while (!std::atomic_compare_exchange_strong(
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_prefix.h:75]: can you explain why you want to use "strong" here?

File: test/inductor/test_torchinductor.py [hunk 3]
```diff
@@ -3646,14 +3646,37 @@ def fn(a, b):
...
+                torch.index_add(a, 0, b, c, alpha=1),
+                torch.index_add(a, 1, b, c, alpha=-1),
             )
 
+        for ind_dtype in (torch.int64, torch.int64):
+            for inp_dtype in (torch.float32,):
+                # torch.bfloat16, torch.half
+                # disable low precision since non-deterministic parallel reduction have larger impact
+                # and also suffered from not using accumulate data type
+                # https://github.com/pytorch/pytorch/issues/93582
```
Comment (jgong5) [inline test/inductor/test_torchinductor.py:3670]: This test is shared by both CPU and GPU. Consider to disable other dtypes only for CPU?

**Review by jgong5 — COMMENTED** [review]
A more performant implementation could be to avoid atomic_add totally. We sort on the indices and then do the parallelization on the sorted indices to avoid thread contention and also enable vectorization. For now, can we just provide this basic support for bf16 atomic_add for function completeness?

**Comment by EikanWang** [comment]
@zhuhaozhe , may I know if you are still working on this PR?

**Pull request closed without merging**